File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/get_market_analysis.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService


@dataclass
class GetMarketAnalysisAction:
    @property
    def name(self) -> str:
        return "GET_MARKET_ANALYSIS"

    async def validate(self, _message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        return True

    async def handler(
        self,
        _message: Message,
        state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        token = state.get("token", "SOL")
        analysis = await service.get_market_analysis(token)

        text = (
            f"Market Analysis for {analysis.token}:\n"
            f"  Trend: {analysis.trend}\n"
            f"  Support: ${analysis.support:.4f}\n"
            f"  Resistance: ${analysis.resistance:.4f}\n"
            f"  Volume (24h): ${analysis.volume_24h:.0f}\n"
            f"  Recommendation: {analysis.recommendation}"
        )

        return ActionResult(True, text, data=analysis)

File: plugins/plugin-auto-trader/rust/src/actions/get_market_analysis.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::{Action, ActionExample, ActionResult};

pub struct GetMarketAnalysisAction;

#[async_trait]
impl Action for GetMarketAnalysisAction {
    fn name(&self) -> &str {
        "GET_MARKET_ANALYSIS"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "MARKET_ANALYSIS",
            "ANALYZE_MARKET",
            "MARKET_REPORT",
            "CHECK_MARKET",
        ]
    }

    fn description(&self) -> &str {
        "Get market analysis for a specific token."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        _message: &Value,
        state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let token = state
            .get("token")
            .and_then(|v| v.as_str())
            .unwrap_or("SOL");

        let analysis = svc.get_market_analysis(token).await;

        let text = format!(
            "Market Analysis for {}:\n  Trend: {}\n  Support: ${:.4}\n  Resistance: ${:.4}\n  Volume (24h): ${:.0}\n  Recommendation: {}",
            analysis.token,
            analysis.trend,
            analysis.support,
            analysis.resistance,
            analysis.volume_24h,
            analysis.recommendation,
        );

        ActionResult {
            success: true,
            text,
            data: Some(json!(analysis)),
            error: None,
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "Analyze the SOL market".to_string(),
            agent_response: "Market Analysis for SOL: Trend: Bullish ...".to_string(),
        }]
    }
}
